Silent component logging

pmuens opened the pull request:
This PR adds the change that the `context.log` function is replaced with `context.debug` before it's passed into components.

With this change verbose component outputs are hidden by default and only visible if one runs the component command with the help of the `--debug` flag (such as `components deploy --debug`). All system logs are still shown by default.

Additionally other log level were implemented which are accessible via `context.console.*` (they follow the JavaScript `console.*` pattern). This means that e.g. component authors can use `context.console.error` to show an error or `context.console.log` to force a log message to be shown.

Here's an example which shows how this can be used as a component author:

```js
const NestedService = {
  shouldDeploy(prevInstance) {
    return 'deploy'
  },

  deploy(prevInstance, context) {
    context.log('Hello World!') // hidden by default (only visible when passing in the --debug flag)
    context.console.error('An error oruccred')
  }
}

module.exports = NestedService
```

/cc @brianneisler

pmuens commented:
I just changed the implementation and added several other log levels. The log level can now be passed into the `context.log` function via an options object.

Before passing the context to the component we modify the `log` function so that it's replaced with a log function which sets the default log `level` to `debug` (this can be overwritten by the component author).

@brianneisler let me know if this is what we're looking for and I'll fix the last failing tests (which is a little bit more complicated due to the way we inject the modified `log` function).

brianneisler commented on src/utils/dag/execGraph.js:
```diff
@@ -2,7 +2,18 @@ import { all, append, isEmpty, reduce } from '@serverless/utils'
+    log: (...args) => {
+      // setting the log level to `debug` if non is provided
+      if (!args.slice(-1)[0].level) {
+        ;[].push.call(args, { level: 'debug' })
```
A simpler approach here might be to use the `append` method. Might need some adjustments to utils to support arguments though.

brianneisler commented on src/utils/logging/debug.js:
```diff
@@ -1,8 +1,8 @@
-    // eslint-disable-next-line no-console
-    console.log(...args)
-  }
+  ;[].push.call(args, { level: 'debug' }) // set log level to `debug`
```
Same comment as above.

brianneisler commented:
@pmuens i'm a little concerned about tacking a value onto the end of the log method like this given the variability of the log method. One thing to consider about this approach is making sure that the config object does not end up in the logs themselves.

pmuens commented on src/utils/logging/debug.test.js:
```diff
@@ -17,17 +17,17 @@ describe('#debug()', () => {
   it('should log message using console.log (node 6 and <= 8.9 support)', () => {
     debug({ options: { debug: true } }, msg)
-    expect(global.console.log).toHaveBeenCalledWith(msg)
+    expect(global.console.debug).toHaveBeenCalledWith(msg)
```
Yes, good catch. I remember that bug hunting process a couple of weeks ago (AFAIK @ganeshvlrk brought that up).

I thought that Babel takes care of that but given that this is more low-level (C / C++ code related) it makes sense that Node handles this differently...

pmuens commented on src/utils/dag/execGraph.js:
```diff
@@ -2,7 +2,18 @@ import { all, append, isEmpty, reduce } from '@serverless/utils'
+    // replacing the `log` function with one which has a default log level of `debug`
+    log: (...args) => {
+      // setting the log level to `debug` if non is provided
+      if (!args.slice(-1)[0].level) {
```
Love that Ramda syntax. Will update that 👍

pmuens commented:
> @pmuens i'm a little concerned about tacking a value onto the end of the log method like this given the variability of the log method. One thing to consider about this approach is making sure that the config object does not end up in the logs themselves.

Thanks for the review @brianneisler 👍
I'll rebase and update this PR based on your feedback.

TBH I'm also not super convinced, that this is the best way to handle this. However I faced a couple of issues while implementing this functionality.

My main goal was to introduce an easy way for component authors to overwrite the default log level (that's why I introduced the `options` object). In my first implementation I attached the log level to the `context`. However this introduced other issues such as global state issues (even if I create a `newContext`) and problems directing the log outputs to the correct log methods when relying on `context.logLevel` (mainly due to the fact that `context.log` is a closure).

Do you think we should implement this differently? Should we keep the `options` object?

pmuens commented:
@brianneisler I just rebased and updated the PR.

I threw away the old implementation and simplified the codebase by introducing new log level methods which can be accessed via `context.console.*`. So if someone wants to get fance he can use `context.console.warn` to show a warning. The normal `context.log` method is replaced with `context.debug` before it's passed down to the components, meaning that normal `context.log` logs are hidden by default and only visible if someone uses the `--debug` flag.
